### ISSUE TITLE
Problem on MacOS

### ISSUE THREAD

**Issue by willmcgugan** [P403n1x87/austin#101]
### Description

I'm unable to get Austin (3.2.0) to run on MacOS 12.0.1

I've installed it via brew. And running it as follows:

```
sudo austin python example.py
```

This is in a venv within my home folder.

Austin reports:

```

🔒 Insufficient permissions. Austin requires the use of sudo on Mac OS in order
to read the memory of even its child processes. Furthermore, the System
Integrity Protection prevents Austin from working with Python binaries
installed in certain areas of the file system. See

    🌐 https://github.com/P403n1x87/austin#compatibility

for more details.
```

Happy to help with testing.

**Comment by P403n1x87** [comment]
@willmcgugan thanks for reporting this. I wonder if the script that you're trying to run terminates too quickly :thinking:. Does it work with this other script?

https://github.com/P403n1x87/austin/blob/904a39075936d18cef8cb5c7d3a688dcc87a4b40/test/target34.py#L25-L38

May I also check with you the Python interpreter version and how it was installed?

**Comment by willmcgugan** [comment]
Afraid not, same error with that script.

**Comment by willmcgugan** [comment]
Python 3.10 Installer from python.org, but running in a venv.

**Comment by P403n1x87** [comment]
Thanks for the extra details. I normally get Python versions installed with pyenv. I'll test with the official installer to see if I can reproduce the problem.

**Comment by P403n1x87** [comment]
I can reproduce the problem with the Python from python.org. The issue is with the venv, which uses symlinks to the binaries:

~~~ console
❯ python3.10 -m venv /tmp/a

❯ ls -l /tmp/a/bin
total 36
-rw-rw-r-- 1 gabriele.tornetta wheel 9033 Dec 18 17:54 Activate.ps1
-rw-rw-r-- 1 gabriele.tornetta wheel 1967 Dec 18 17:54 activate
-rw-rw-r-- 1 gabriele.tornetta wheel  893 Dec 18 17:54 activate.csh
-rw-rw-r-- 1 gabriele.tornetta wheel 2035 Dec 18 17:54 activate.fish
-rwxr-xr-x 1 gabriele.tornetta wheel  234 Dec 18 17:54 pip
-rwxr-xr-x 1 gabriele.tornetta wheel  234 Dec 18 17:54 pip3
-rwxr-xr-x 1 gabriele.tornetta wheel  234 Dec 18 17:54 pip3.10
lrwxr-xr-x 1 gabriele.tornetta wheel   10 Dec 18 17:54 python -> python3.10
lrwxr-xr-x 1 gabriele.tornetta wheel   10 Dec 18 17:54 python3 -> python3.10
lrwxr-xr-x 1 gabriele.tornetta wheel   25 Dec 18 17:54 python3.10 -> /usr/local/bin/python3.10
~~~

Copying over the python3.10 binary might not "fix" this, since the binary still links to a shared library in a protected area of the file system:

~~~ console
❯ otool -L `which python`
/tmp/a/bin/python:
	/Library/Frameworks/Python.framework/Versions/3.10/Python (compatibility version 3.10.0, current version 3.10.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
~~~

Other installation methods don't suffer from this issue because they install the binaries in non-protected locations:

~~~ console
❯ python3.9 -m venv /tmp/b

❯ ls -l /tmp/b/bin
total 36
-rw-r--r-- 1 gabriele.tornetta wheel 8834 Dec 18 17:57 Activate.ps1
-rw-r--r-- 1 gabriele.tornetta wheel 1878 Dec 18 17:57 activate
-rw-r--r-- 1 gabriele.tornetta wheel  827 Dec 18 17:57 activate.csh
-rw-r--r-- 1 gabriele.tornetta wheel 1967 Dec 18 17:57 activate.fish
-rwxr-xr-x 1 gabriele.tornetta wheel  233 Dec 18 17:57 pip
-rwxr-xr-x 1 gabriele.tornetta wheel  233 Dec 18 17:57 pip3
-rwxr-xr-x 1 gabriele.tornetta wheel  233 Dec 18 17:57 pip3.9
lrwxr-xr-x 1 gabriele.tornetta wheel    9 Dec 18 17:57 python -> python3.9
lrwxr-xr-x 1 gabriele.tornetta wheel    9 Dec 18 17:57 python3 -> python3.9
lrwxr-xr-x 1 gabriele.tornetta wheel   39 Dec 18 17:57 python3.9 -> /usr/local/opt/python@3.9/bin/python3.9

❯ otool -L `which python3.9`
/usr/local/bin/python3.9:
	/usr/local/Cellar/python@3.9/3.9.7/Frameworks/Python.framework/Versions/3.9/Python (compatibility version 3.9.0, current version 3.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
~~~

Unfortunately Austin has no way of working around this, unless the security policies are changed manually by the user 🙁 .

**Comment by P403n1x87** [comment]
@willmcgugan granted there is nothing I can do about this but persuade you to install Python either via `brew` or `pyenv`, I'd suggest we close this issue. WDYT?

**Comment by willmcgugan** [comment]
Sounds like that's my only option. Thanks for looking in to it!

**Comment by willmcgugan** [comment]
Hi, I've been trying to use Austin again on MacOS. I have `pyenv` installed and working in a venv. But every invocation of Austin results in "insufficient permissions".

This is on an M1 Mac if that makes a difference.

**Comment by P403n1x87** [comment]
@willmcgugan just to rule out the obvious, as this seems to still be the same issue: can I check with you that the venv was created with a version of Python installed with `pyenv`? Also, what is the Python version? Could you share the output of

~~~ console
otool -L `which python`
~~~

run within the venv, please? I presume this time it's Austin 3.3?

**Comment by willmcgugan** [comment]
Yes, it was a venv with Python installed with pyenv. Python version is 3.10.2

The command gives me this:

```
(textual) willmcgugan@Wills-MacBook-Pro sandbox % otool -L `which python`
/Users/willmcgugan/.pyenv/shims/python: is not an object file
```

Austin version is 3.3.0

**Comment by P403n1x87** [comment]
~~~
/Users/willmcgugan/.pyenv/shims/python: is not an object file
~~~

Hmm that's weird. Within a venv I'd expect `which python` to return the location of the Python binary within the venv, i.e.
`$VIRTUAL_ENV/bin/python`. I suspect `/Users/willmcgugan/.pyenv/shims/python` is actually pointing to the system-installed Python, whence the issue.

**Comment by willmcgugan** [comment]
That actually works!

Not sure why `which python` points at the shim. I'm using the pyenv extension for virtualenvs, which seems to work fine. So the shim must be setting the venv environment.

**Comment by P403n1x87** [comment]
Thanks for letting me know. I will then close this issue again for now since the follow-up seems to be related to pyenv.

**Comment by tonybaloney** [comment]
@P403n1x87 this should really be in the docs if its the only way to get it working on macOS

**Comment by P403n1x87** [comment]
> @P403n1x87 this should really be in the docs if its the only way to get it working on macOS

@tonybaloney I sneaked a change in the README with the 3.3.0 release

https://github.com/P403n1x87/austin/commit/c0765082320ce5a365fd74dcc5a6c29b54a3c87e#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R563-R566

Do you reckon there is something else worth mentioning in that note, or elsewhere? As a `pyenv` user on macOS I don't really know what other users are experiencing with their preferred setup so I'm very keen to hear other's experiences on this to improve the docs.

**Comment by tonybaloney** [comment]
> > @P403n1x87 this should really be in the docs if its the only way to get it working on macOS
> 
> 
> 
> @tonybaloney I sneaked a change in the README with the 3.3.0 release, in response to the original issue:
> 
> 
> 
> https://github.com/P403n1x87/austin/commit/c0765082320ce5a365fd74dcc5a6c29b54a3c87e#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R563-R566
> 
> 
> 
> Do you reckon there is something else worth mentioning in that note, or elsewhere? As a `pyenv` user on macOS I don't really know what other users are experiencing with their preferred setup so I'm very keen to hear other's experiences on this to improve the docs.

IMO the link in the error message should go to a unique section for macOS users. 
I've never managed to get Austin working on an "official" version of Python, in or outside a virtual environment. 
I can only use it on binaries that were compiled on my machine. Pyenv is the most convenient way of doing that. 

The root user thing is a red herring to first time users.
